### PR TITLE
Option to limit the number of rounds of enumerative instantiation

### PR DESCRIPTION
**Pull request by ajreynol**
Possible fix for #3481.

### CODE REVIEW

File: src/theory/quantifiers/inst_strategy_enumerative.cpp [hunk 1]
```diff
@@ -33,12 +33,19 @@ using namespace inst;
...
 {
 }
-
+void InstStrategyEnum::presolve()
+{
+  d_fullSaturateLimit = options::fullSaturateLimit();
+}
 bool InstStrategyEnum::needsCheck(Theory::Effort e)
 {
+  if (d_fullSaturateLimit == 0)
```
Comment (HanielB) [inline src/theory/quantifiers/inst_strategy_enumerative.cpp]: shouldn't this be checking `<= 0`?

File: src/theory/quantifiers/inst_strategy_enumerative.cpp [hunk 2]
```diff
@@ -61,15 +68,18 @@ void InstStrategyEnum::check(Theory::Effort e, QEffort quant_e)
 {
   bool doCheck = false;
   bool fullEffort = false;
-  if (options::fullSaturateInterleave())
+  if (d_fullSaturateLimit != 0)
```
Comment (HanielB) [inline src/theory/quantifiers/inst_strategy_enumerative.cpp]: shouldn't this be checking `> 0`?

File: src/theory/quantifiers/inst_strategy_enumerative.cpp [hunk 3]
```diff
@@ -150,6 +160,10 @@ void InstStrategyEnum::check(Theory::Effort e, QEffort quant_e)
     Trace("fs-engine") << "Finished full saturation engine, time = "
                        << (clSet2 - clSet) << std::endl;
   }
+  if (d_fullSaturateLimit > 0)
+  {
+    d_fullSaturateLimit--;
+  }
```
[the remark this answers is not in the document]
Comment (ajreynol) [reply src/theory/quantifiers/inst_strategy_enumerative.cpp]: 0 = no more rounds, -1 = unlimited rounds

File: src/theory/quantifiers/inst_strategy_enumerative.cpp (hunk 1 above)
Comment (ajreynol) [reply src/theory/quantifiers/inst_strategy_enumerative.cpp]: No, -1 means unlimited.

File: src/theory/quantifiers/inst_strategy_enumerative.cpp (hunk 2 above)
Comment (ajreynol) [reply src/theory/quantifiers/inst_strategy_enumerative.cpp]: Ditto.

File: src/theory/quantifiers/inst_strategy_enumerative.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (HanielB) [reply src/theory/quantifiers/inst_strategy_enumerative.cpp]: Got it mixed up, sorry. But the test below with `!= 0` would pass with any negative number, so it does not matter if it's -1 or less, right?

[the remark this answers is not in the document]
Comment (ajreynol) [reply src/theory/quantifiers/inst_strategy_enumerative.cpp]: Yes, but it'd be pointless to decrement when <0.

[the remark this answers is not in the document]
Comment (HanielB) [reply src/theory/quantifiers/inst_strategy_enumerative.cpp]: It'd make the code simpler.

[the remark this answers is not in the document]
Comment (ajreynol) [reply src/theory/quantifiers/inst_strategy_enumerative.cpp]: It makes it possible to underflow and does needless work.

[the remark this answers is not in the document]
Comment (ajreynol) [reply src/theory/quantifiers/inst_strategy_enumerative.cpp]: It also makes the invariant about the value of this variable more convoluted.

[the remark this answers is not in the document]
Comment (HanielB) [reply src/theory/quantifiers/inst_strategy_enumerative.cpp]: All right, let's leave it like that.

**Pull request merged**